pkg/url/url.go

```diff
@@ -512,16 +542,46 @@ func ListIngressAndRoute(oclient *occlient.Client, client *kclient.Client, envSp
 		clusterURL := getMachineReadableFormat(r)
 		clusterURLMap[clusterURL.Name] = clusterURL
 	}
-	for _, envinfoURL := range localEnvinfoURLs {
-		// only checks for Ingress and Route URLs
-		if envinfoURL.Kind == envinfo.DOCKER {
-			continue
-		}
-		if !routeSupported && envinfoURL.Kind == envinfo.ROUTE {
-			continue
+	for _, endpointMap := range containerEndpointsMap {
+		for _, localEndpoint := range endpointMap {
+			// only exposed endpoint will be shown as a URL in `odo url list`
+			if localEndpoint.Exposure == common.None || localEndpoint.Exposure == common.Internal {
+				continue
+			}
+			var devfileURL envinfo.EnvInfoURL
+			for _, envinfoURL := range localEnvinfoURLs {
+				// only checks for Ingress and Route URLs
+				if envinfoURL.Name != localEndpoint.Name {
+					continue
+				}
+				if envinfoURL.Kind == envinfo.DOCKER {
+					continue
+				}
+				if !routeSupported && envinfoURL.Kind == envinfo.ROUTE {
+					continue
+				}
+				devfileURL = envinfoURL
+				devfileURL.Port = int(localEndpoint.TargetPort)
+				devfileURL.Secure = localEndpoint.Secure
+			}
+			if reflect.DeepEqual(devfileURL, envinfo.EnvInfoURL{}) {
+				// Devfile endpoint by default should create a route if no host information is provided in env.yaml
+				// If it is not openshify cluster, should ignore the endpoint entry when executing url describe/list
+				if !routeSupported {
+					break
+				}
+				devfileURL.Name = localEndpoint.Name
+				devfileURL.Port = int(localEndpoint.TargetPort)
+				devfileURL.Secure = localEndpoint.Secure
+				devfileURL.Kind = envinfo.ROUTE
+			}
+			localURL := ConvertEnvinfoURL(devfileURL, componentName)
+			// use the trimmed URL Name as the key since remote URLs' names are trimmed
+			trimmedURLName := strings.TrimSpace(util.GetDNS1123Name(strings.ToLower(localURL.Name)))
+			trimmedURLName = util.TruncateString(trimmedURLName, 15)
```

Comment:
These steps are present in GetIngressOrRoute() too. Please put them in a function to maintain consistency.

Reply:
updated